### DISCUSSION TITLE
Hide / Remove bounding Box at Inference

### DISCUSSION THREAD

**Discussion by hacadtra** [facebookresearch/detectron2#2992] [Q&A] [answered]
Hi,
Is there a way to not display bounding box and class name when inferencing model?
In my case, I will only count the total detection and show only the instance segmentation of the given image. I only use 1 class for this detection.

normal result
![normal inference](https://user-images.githubusercontent.com/81913439/116658720-9db79f00-a9ba-11eb-92bb-0ab427501d76.png)

desired result
![desired inference result](https://user-images.githubusercontent.com/81913439/116658776-b32cc900-a9ba-11eb-8311-8e2371d2fe87.jpeg)

for the normal result I've used this code:
```
im = cv2.imread(d["file_name"])
outputs = predictor(I'm)
v = Visualizer(im[:, :, ::-1], metadata=microcontroller_metadata, scale=0.8)
v = v.draw_instance_predictions(outputs["instances"].to("cpu"))
plt.figure(figsize = (14, 10))
plt.imshow(cv2.cvtColor(v.get_image()[:, :, ::-1], cv2.COLOR_BGR2RGB))
plt.show()
```

hope you can give some suggestions

**Comment by VGrondin** [comment]
Yes you can, do you still need it?

**Reply by hacadtra** [reply]
yes, please enlighten me

**Comment by VGrondin** [comment, marked as answer]
You can use overlay_instances. Here is a code snippet:
```
# visualize
im = cv2.imread(d["file_name"])
# predictions
outputs_pred = predictor(im)   
v = Visualizer(img,
                metadata=metadata, 
                scale=1, 
                instance_mode =  ColorMode.BW     # remove color from image, better see segmentation mask  
)
dt_image = v.overlay_instances(
                boxes=None, 
                masks=outputs_pred["instances"].pred_masks.to("cpu"),  
)
        
cv2.imshow('predictions', dt_image[:, :, ::-1])
cv2.waitKey(0)
```

If your image colors are reversed, remove the [:, :, ::-1]

**Reply by hacadtra** [reply]
such a relief, thanks @VGrondin

**Reply by sushmasuresh28** [reply]
Thank you for the solution. I found it very helpful!

@VGrondin Is there a way to just get the bounding boxes with the prediction percentage but no class names or custom class names?

Since I am using a pre-trained model (which I am saving after detectron training in another file), I do not have the metadata registered while making the predictions

**Reply by VGrondin** [reply]
@sushmasuresh28 Hi I know for sure you can do it with custom class name, not sure what it will do if you use no class name. Here is a snippet:
```
from detectron2.utils.visualizer import Visualizer, _create_text_labels

metadata = MetadataCatalog.get("YOUR_DATASET_NAME")      # the dataset you trained you networl on. Same name you used cfg.DATASETS.TRAIN = ("YOUR_DATASET_NAME",) 

# custom labels 
labels = _create_text_labels(outputs_pred["instances"].pred_classes.to("cpu"), outputs_pred["instances"].scores.to("cpu"), metadata.thing_classes) 

#  retake the code from accepted answer
dt_image = v.overlay_instances(
                boxes=boxes=outputs_pred["instances"].pred_boxes.to("cpu"), 
                labels=labels,
                masks=outputs_pred["instances"].pred_masks.to("cpu"),  
                assigned_colors=None
)
```

If you don't have metadata.thing_classes, you can manually register it with something like:
```
MetadataCatalog.get("YOUR_DATASET_NAME").set(thing_classes=["class_01", "class_02"]
metadata = MetadataCatalog.get("YOUR_DATASET_NAME") 
```

Note you can also predefine colors! Hope it helps

**Reply by sushmasuresh28** [reply]
@VGrondin this totally helped! Thank you so much :)
 
With no class name, the results are as follows-
![image](https://user-images.githubusercontent.com/58828813/125894679-793d0ce5-fb48-4991-8fff-3e65c957bf44.png)

After I manually register(as you suggested), I get exactly what I need-

![image](https://user-images.githubusercontent.com/58828813/125894989-4e3409bb-097d-4ce6-a70c-4b23991bf031.png)

Out of curiosity, I am trying to compute the area enclosed by each of the predicted polygons. I have been looking to access the coordinates of the drawn polygons but can't seem to find how they are stored.

By running `print(outputs["instances"].pred_boxes)` I get a list of tensors like this -
![image](https://user-images.githubusercontent.com/58828813/125894092-8b9c6ee2-aaca-4c44-905b-6775d2502e68.png)
Do you know what this information would actually mean?

The documentation here - https://detectron2.readthedocs.io/en/latest/tutorials/models.html suggests that the "instance" object has"pred_masks" which should contain  "a Tensor of shape (N, H, W), masks for each detected instance".

But when I run `print(outputs["instances"].pred_masks)` I get -
![image](https://user-images.githubusercontent.com/58828813/125894339-657fb9c2-e9eb-4b1c-bf95-1bc538dafa0b.png). Not sure why this is the case when clearly I am able to see the masks on the image.

I don't understand how to obtain the polygon information to calculate the area. If you know how the information can be accessed, could you please help me?

**Reply by VGrondin** [reply]
Glad it worked! 

The  mask format you printed is a binary image. I guess it the fastest method for models during inference time, otherwise it would have to encode the binary image into polygon/rle format which would slow down the process. Anyways binary images are easy, to handle:
```
# print number of true pixels (area) for each i predicted instances
for i in range(len(outputs["instances"].pred_masks)):
    print(torch.sum(outputs["instances"].pred_masks[i]))
```

**Reply by sushmasuresh28** [reply]
@VGrondin This totally makes sense! Thank you for clarifying :)

I just have one final question that I can't seem to find the answer to that anywhere(including the documentation [here](https://detectron2.readthedocs.io/en/latest/tutorials/models.html)). 

As shown above, when I run  `print(outputs["instances"].pred_boxes)` I get a list of tensors like this -
![image](https://user-images.githubusercontent.com/58828813/125894092-8b9c6ee2-aaca-4c44-905b-6775d2502e68.png)

Are the printed tensors representing the edge coordinates of each predicted bounding box? If yes, why are they float values/ decimals? If they are not the bounding box points, then what exactly are they? 
If you happen to know what `pred_boxes` is actually representing and can share that with me, it will surely make my understanding of this concept better

**Reply by IAmOZRules** [reply]
hi
@VGrondin  can you pls suggest how to get the coordinates of the polygon. As you suggested torch.sum gives the area but for my application, I need the X,Y coordinates of each of the polygon points. Thanks

**Reply by VGrondin** [reply]
```
# convert binary mask to polygon mask
has_mask = instances.has("pred_masks")
if has_mask:
     segmentations = []
     areas = []
     for mask in instances.pred_masks:
         polygons = Mask(mask.numpy()).polygons()
         segmentations.append(polygons.segmentation)
         
         area = Mask(mask.numpy()).area()
         areas.append(int(area))`
